### ISSUE TITLE
[ENH] GRAN-DAG Algorithm

### ISSUE THREAD

**Issue by Manas-7854** [pgmpy/pgmpy#3405]
**Is your feature request related to a problem? Please describe.**
 GraN-DAG (Gradient-based Neural DAG Learning) parameterizes each node’s conditional distribution using a neural network that takes all other variables as input. A continuous differentiable acyclicity constraint (adapted from NOTEARS) is imposed so that the entire structure learning problem can be solved end-to-end with gradient descent. A Lagrangian augmentation scheme is used to enforce the acyclicity constraint as a hard constraint at convergence. Unlike linear NOTEARS, GraN-DAG captures non-linear relationships without requiring explicit functional form assumptions.

**Additional context**
Paper: [Gradient-Based Neural DAG Learning](https://arxiv.org/abs/1906.02226)
Reference codebase: [kurowasan/GraN-DAG](https://github.com/kurowasan/GraN-DAG)
Related : #2931, pgmpy/enhancement_proposals#4

Note that this is part of an ongoing mentorship project
